### PR TITLE
v4 upgrade

### PR DESCRIPTION
**Pull request by asalzmann**
Co-authored-by: Hieu Vu <72878483+hieuvubk@users.noreply.github.com>
Co-authored-by: Jacob Gadikian <jacobgadikian@gmail.com>

Closes: #384

## Context and purpose of the change

As part of the [v3](https://github.com/Stride-Labs/stride/releases/tag/v3.0.0) upgrade, we added [authz](https://github.com/Stride-Labs/stride/pull/352). However, we forgot to set the store key as part of the upgrade. As a consequence, after the upgrade, validators that attempted to prune had apphash errors. More details from @chillyvee [here](https://github.com/cosmos/cosmos-sdk/issues/13933#issuecomment-1326791953).

## Brief Changelog

- Add upgrade_tests to the v3 / v4 upgrade
- Initialize the params store for the claim and authz modules
- Add v4 upgrade handler and set the `authzkeeper.StoreKey`

## Author's Checklist

I have...

- [ ] Run and PASSED locally all GAIA integration tests
- [x] If the change is contentful, I either:
    - [x] Added a new unit test OR 
    - [ ] Added test cases to existing unit tests
- [ ] OR this change is a trivial rework / code cleanup without any test coverage

If skipped any of the tests above, explain.

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] manually tested (if applicable)
- [ ] confirmed the author wrote unit tests for new logic
- [ ] reviewed documentation exists and is accurate

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes? 
  - [ ] Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`?
  - [ ] This pull request updates existing proto field values (and require a backend and frontend migration)? 
  - [ ] Does this pull request change existing proto field names (and require a frontend migration)?
  How is the feature or change documented? 
      - [ ] not applicable
      - [ ] jira ticket `XXX` 
      - [ ] specification (`x/<module>/spec/`) 
      - [ ] README.md 
      - [ ] not documented

### CODE REVIEW

**Comment by asalzmann** [comment]
I couldn't push to the other branch (https://github.com/Stride-Labs/stride/pull/384) to resolve the linting errors, so created a new one here. There are no changes aside from merging main into the original branch.

**Comment by sampocs** [comment]
See comments in #384
Also, can we add the [param store key ](https://github.com/Stride-Labs/stride/pull/393/files#diff-aeec7ddb4cf8b7e9e6e57625770d369cce5f377cc350b42c3aca0427bba8dfbcR52-R55) to the instructions in`upgrades/README.md`

**Review by faddat — APPROVED** [review]

**Comment by faddat** [comment]
Others should have a look also note that I merged in main.

**Comment by sampocs** [comment]
@asalzmann this diverged from #384 - wanna merge in the recent changes into this one?

**Comment by asalzmann** [comment]
Merged in. @faddat is it possible for me to get access to the notional stride fork? It would simplify the collaboration workflow (right now my workflow is to merge the notional branch into a separate stride branch, then that stride branch into main, which is a bit cumbersome)

File: app/apptesting/test_helpers.go [hunk 1]
```diff
@@ -36,6 +39,7 @@ var (
 type AppTestHelper struct {
 	suite.Suite
 
+	Context 	sdk.Context
```
Comment (sampocs) [inline app/apptesting/test_helpers.go:42]: nit: rm

File: app/apptesting/test_helpers.go [hunk 2]
```diff
@@ -57,6 +61,7 @@ func (s *AppTestHelper) Setup() {
 		GRPCQueryRouter: s.App.GRPCQueryRouter(),
 		Ctx:             s.Ctx(),
 	}
+	s.Context = s.App.BaseApp.NewContext(false, tmtypes.Header{Height: 1, ChainID: StrideChainID, Time: time.Now().UTC()})
```
Comment (sampocs) [inline app/apptesting/test_helpers.go:64]: nit: rm

**Review by sampocs — COMMENTED** [review]
lgtm! Two nit comments

**Comment by faddat** [comment]
Yes, I'll add you there.

**Comment by faddat** [comment]
@asalzmann there's an invite for you at:

https://github.com/notional-labs/stride/invitations

**Review by faddat — CHANGES_REQUESTED** [review]
This PR should change the import path, I think.  If we'd like to do that separately, please let me know, and I'll un-request these changes :)

**Comment by sampocs** [comment]
I tested this upgrade locally and it's erroring when trying to add the new authz store key. After removing the authz store upgrade snippet (`Added: []string{authzkeeper.StoreKey}`), the upgrade was successful.

It also appears that the authz store key already exists. I printed `app.keys` from the upgrade handler and `authz` was in the map.

**Comment by sampocs** [comment]
Also tested it locally using the mainnet state and got the same error
```
state_export-stride-1  | failed to load latest version: failed to load store: initial version set to 1234450, but found earlier version 1234127
```

Removing the authz store upgrade snippet solved this as well.

**Comment by sampocs** [comment]
@asalzmann was the renaming of `Stride-Labs` → `Stride-labs` intentional?

**Comment by sampocs** [comment]
I reverted the v4 import path updates - let's do that in a separate PR as we continue to debug this handler. @faddat can you remove the change request?

**Review by faddat — APPROVED** [review]

File: app/app.go [hunk 3]
```diff
@@ -1002,6 +1002,8 @@ func initParamsKeeper(appCodec codec.BinaryCodec, legacyAmino *codec.LegacyAmino
 	paramsKeeper.Subspace(icacallbacksmoduletypes.ModuleName)
 	// this line is used by starport scaffolding # stargate/app/paramSubspace
 
+	paramsKeeper.Subspace(claimtypes.ModuleName)
+	paramsKeeper.Subspace(authz.ModuleName)
```
Comment (asalzmann) [inline app/app.go:1006]: Let's leave this out of this upgrade

File: app/upgrades.go [hunk 4]
```diff
@@ -41,9 +77,8 @@ func (app *StrideApp) setupUpgradeHandlers() {
 		storeUpgrades = &storetypes.StoreUpgrades{
 			Added: []string{claimtypes.StoreKey},
 		}
-	}
-
-	if storeUpgrades != nil {
 		app.SetStoreLoader(upgradetypes.UpgradeStoreLoader(upgradeInfo.Height, storeUpgrades))
+	case "v4":
+		app.SetStoreLoader(AuthzHeightAdjustmentUpgradeStoreLoader(upgradeInfo.Height))
```
Comment (asalzmann) [inline app/upgrades.go:82]: Let's move this to v5

File: app/upgrades/v4/upgrades_test.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+			func() {
+				suite.Setup()
+			},
+			func() {
+				suite.ConfirmUpgradeSucceededs("v4", dummyUpgradeHeight)
+
+				// make sure authz module was init
+				afterCtx := suite.Ctx().WithBlockHeight(dummyUpgradeHeight)
+				actGenState := suite.App.AuthzKeeper.ExportGenesis(afterCtx)
+				expGenState := authz.DefaultGenesisState()
```
Comment (asalzmann) [inline app/upgrades/v4/upgrades_test.go:47]: rm

Comment (sampocs) [reply app/upgrades/v4/upgrades_test.go]: done

File: app/upgrades.go (hunk 4 above)
Comment (sampocs) [reply app/upgrades.go]: done

File: app/upgrades/v4/README.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# Upgrade v4 Changelog
```
[the remark this answers is not in the document]
Comment (sampocs) [reply app/upgrades/v4/README.md:1]: chatted, leaving this in

**Pull request merged**